impeller/geometry/path.h

```diff
@@ -75,107 +77,261 @@ class Path {
+  }
 
   bool GetQuadraticComponentAtIndex(size_t index,
-                                    QuadraticPathComponent& quadratic) const;
+                                    QuadraticPathComponent& quadratic) const {
+    return state_->GetQuadraticComponentAtIndex(index, quadratic);
+  }
 
-  bool GetCubicComponentAtIndex(size_t index, CubicPathComponent& cubic) const;
+  bool GetCubicComponentAtIndex(size_t index, CubicPathComponent& cubic) const {
+    return state_->GetCubicComponentAtIndex(index, cubic);
+  }
 
   bool GetContourComponentAtIndex(size_t index,
-                                  ContourComponent& contour) const;
+                                  ContourComponent& contour) const {
+    return state_->GetContourComponentAtIndex(index, contour);
+  }
 
   bool UpdateLinearComponentAtIndex(size_t index,
-                                    const LinearPathComponent& linear);
+                                    const LinearPathComponent& linear) {
+    return state_->UpdateLinearComponentAtIndex(index, linear);
+  }
 
-  bool UpdateQuadraticComponentAtIndex(size_t index,
-                                       const QuadraticPathComponent& quadratic);
+  bool UpdateQuadraticComponentAtIndex(
+      size_t index,
+      const QuadraticPathComponent& quadratic) {
+    return state_->UpdateQuadraticComponentAtIndex(index, quadratic);
+  }
 
-  bool UpdateCubicComponentAtIndex(size_t index, CubicPathComponent& cubic);
+  bool UpdateCubicComponentAtIndex(size_t index, CubicPathComponent& cubic) {
+    return state_->UpdateCubicComponentAtIndex(index, cubic);
+  }
 
   bool UpdateContourComponentAtIndex(size_t index,
-                                     const ContourComponent& contour);
+                                     const ContourComponent& contour) {
+    return state_->UpdateContourComponentAtIndex(index, contour);
+  }
 
   /// Callers must provide the scale factor for how this path will be
   /// transformed.
   ///
   /// It is suitable to use the max basis length of the matrix used to transform
   /// the path. If the provided scale is 0, curves will revert to lines.
-  Polyline CreatePolyline(Scalar scale) const;
+  Polyline CreatePolyline(Scalar scale) const {
+    return state_->CreatePolyline(scale);
+  }
 
-  std::optional<Rect> GetBoundingBox() const;
+  std::optional<Rect> GetBoundingBox() const {
+    return state_->GetBoundingBox();
+  }
 
-  std::optional<Rect> GetTransformedBoundingBox(const Matrix& transform) const;
+  std::optional<Rect> GetTransformedBoundingBox(const Matrix& transform) const {
+    return state_->GetTransformedBoundingBox(transform);
+  }
 
-  std::optional<std::pair<Point, Point>> GetMinMaxCoveragePoints() const;
+  std::optional<std::pair<Point, Point>> GetMinMaxCoveragePoints() const {
+    return state_->GetMinMaxCoveragePoints();
+  }
 
  private:
   friend class PathBuilder;
 
-  void SetConvexity(Convexity value);
+  void SetConvexity(Convexity value) { state_->SetConvexity(value); }
```

Comment:
I think its pretty confusing to have two different classes called state in the same header. Why do we need to move all of the path state into this object?

Reply:
Because path caches polyline and polyline caches tesselation data. With this in place we now care about what a path instance actually is. So the alternatives would be to make path non copyable and use shared_ptrs everywhere, or make path non copyable, have one owner and pass by reference everywhere (that would become problematic the moment we want to pass it to std::function for example).

Making the path state internally reference seemed like the least disruptive change. But it's possible least disruptive is not what the objective is :)

Reply:
I don't understand what the value is of having two separate caches. Seems like Polyline data is almost always immediately passed to either Tessellate or Tessellate convex. And anything that invalidates the tessellated data _always_ invalidates the polyline too. Can use just cache the tessellation output?